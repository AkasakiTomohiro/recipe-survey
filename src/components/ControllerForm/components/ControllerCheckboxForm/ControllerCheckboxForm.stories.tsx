/* eslint-disable react-hooks/rules-of-hooks */
/**
 * @file ControllerCheckboxForm.stories.tsx
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import { FormProvider, useForm } from 'react-hook-form';

import { ControllerCheckboxForm } from './ControllerCheckboxForm';

import type { Meta, StoryObj } from '@storybook/react';

type Test = {
  name: boolean
}

const meta: Meta<typeof ControllerCheckboxForm> = {
  title : 'RecipeSurvey/components/ControllerForm/ControllerCheckboxForm',
  render: (args) => {
    const methods = useForm<Test>({
      defaultValues: {
        name: true
      }
    });

    return (
      <FormProvider {...methods} >
        <ControllerCheckboxForm {...args} />
      </FormProvider>
    );
  }
};

export default meta;

type Story = StoryObj<typeof ControllerCheckboxForm>;

export const Primary: Story = {
  args: {
    label: 'チェックボックス',
    name : 'name'
  }
};
