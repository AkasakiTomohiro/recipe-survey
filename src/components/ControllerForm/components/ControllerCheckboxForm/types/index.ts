/**
 * @file index.ts
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import { FieldPath } from 'react-hook-form';

import { RequiredFilterType } from '@/types';

export interface IControllerCheckboxFormProps<T extends Record<string, any> = Record<string, any>> {

  /**
   * フォームの名前
   */
  label?: string;

  /**
   * controlで管理しているオブジェクトのキーを指定
   */
  name: FieldPath<RequiredFilterType<T, boolean>>;

  /**
   * Grid幅を横全体にするか
   */
  allGird?: boolean;
}