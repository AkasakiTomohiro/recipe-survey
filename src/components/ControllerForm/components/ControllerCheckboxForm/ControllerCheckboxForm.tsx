/**
 * @file ControllerCheckboxForm.tsx
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import { useController, useFormContext } from 'react-hook-form';

import { Checkbox, FormControlLabel } from '@mui/material';

import { GridFormControl } from '../GridFormControl/GridFormControl';
import { IControllerCheckboxFormProps } from './types';

/**
 * Primary UI component for user interaction
 *
 * @example
 *
 * 利用したいコンポーネントでジェネリックに型を定義する形でラップして利用すること
 *
 * const CheckboxForm = (props: IControllerCheckboxFormProps<XXXType>) => {
 *   return (<ControllerCheckboxForm {...props}/>);
 * };
 */
export const ControllerCheckboxForm = ({
  label,
  name,
  isAllGird,
  ...otherProps
}: IControllerCheckboxFormProps): JSX.Element => {
  const { control, getValues } = useFormContext();
  const { field } = useController({ name: name, control });
  const defaultValues = getValues();

  const CheckBoxWrap = () => <Checkbox {...otherProps} {...field} defaultChecked={defaultValues[name]} />
  
  if(label === undefined) {
    return (
      <GridFormControl isAllGird={isAllGird}>
        <CheckBoxWrap />
      </GridFormControl>
    );
  }

  return (
    <GridFormControl isAllGird={isAllGird}>
      <FormControlLabel 
        control={<CheckBoxWrap />}
        label={label}
      />
    </GridFormControl>
  );
};
