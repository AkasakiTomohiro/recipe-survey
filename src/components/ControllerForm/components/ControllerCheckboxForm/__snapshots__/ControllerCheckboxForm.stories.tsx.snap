// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ControllerForm/ControllerCheckboxForm NoLabel smoke-test 1`] = `
<div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
  <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
    <span class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-10aivp9-MuiButtonBase-root-MuiCheckbox-root">
      <input class="PrivateSwitchBase-input css-1m9pwf3"
             name="name"
             type="checkbox"
             data-indeterminate="false"
             value="true"
             checked
      >
      <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
           focusable="false"
           aria-hidden="true"
           viewbox="0 0 24 24"
           data-testid="CheckBoxIcon"
      >
        <path d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z">
        </path>
      </svg>
      <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
      </span>
    </span>
  </div>
</div>
`;

exports[`components/ControllerForm/ControllerCheckboxForm Primary smoke-test 1`] = `
<div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
  <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
    <label class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root">
      <span class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-10aivp9-MuiButtonBase-root-MuiCheckbox-root">
        <input class="PrivateSwitchBase-input css-1m9pwf3"
               name="name"
               type="checkbox"
               data-indeterminate="false"
               value="true"
               checked
        >
        <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
             focusable="false"
             aria-hidden="true"
             viewbox="0 0 24 24"
             data-testid="CheckBoxIcon"
        >
          <path d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z">
          </path>
        </svg>
        <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
        </span>
      </span>
      <span class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root">
        チェックボックス
      </span>
    </label>
  </div>
</div>
`;
